First batch of improvements to Carousel

sgomes opened the pull request:
This patch is the first of a series of changes that aim to improve the Carousel and ultimately remove its `jQuery` dependency, while keeping it IE11-compatible.

I'm trying to keep the changes relatively small and self-contained so that they can be incrementally tested (and debugged, if needed), instead of dropping a gigantic change that rewrites the world.

This PR syncs Jetpack to D56414-code, and applies some extra lint fixes, since the rules appear to be different. The change was started on wpcom instead of Jetpack due to legacy reasons. The next batch of improvements works on making that unnecessary.

This PR is a manual sync, since the automated tool appears to fail due to the mismatch in directory structure for the RTL CSS.

#### Changes proposed in this Pull Request:
* Move carousel skeleton HTML generation to the server. It only gets generated once, and it can be made to be device-independent, so there's no point doing it in JS. Note that populating the skeleton with data for a particular gallery is still done in JS, as is some of the resizing logic.
* Remove map generation code from JS.
* Modernise some of the layout to rely more on CSS and less on JS, by e.g. making use of flexbox.
* Rework padding calculations to be simpler and to work consistently across window resizes.
* Don't animate padding-related transitions if the user prefers reduced motion.
* Move some styles away from inlined style attributes, and into the CSS file.
* Remove Date.now polyfill, which is only applicable to ancient browsers.

#### Jetpack product discussion
See p1HpG7-bfN-p2

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
* Smoke-test the carousel and ensure that it continues to work in the exact same way.

#### Proposed changelog entry for your changes:
* Carousel: under-the-hood code improvements.

jeherve commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:
```diff
@@ -357,10 +335,133 @@ function enqueue_assets() {
+									<?php if ( ! $localize_strings['is_logged_in'] && $localize_strings['comment_registration'] ) : ?>
+										<div id="jp-carousel-comment-form-commenting-as">
+											<p id="jp-carousel-commenting-as">
+												<?php // phpcs:ignore Standard.Category.SniffName.ErrorCode,WordPress.Security.EscapeOutput.UnsafePrintingFunction ?>
+												<?php _e( 'You must be <a href="#" class="jp-carousel-comment-login">logged in</a> to post a comment.', 'jetpack' ); ?>
```
Could we do something like this instead?

```suggestion
												<?php
													echo wp_kses(
														__( 'You must be <a href="#" class="jp-carousel-comment-login">logged in</a> to post a comment.', 'jetpack' ),
														array(
															'a' => array(
																'href'  => array(),
																'class' => array(),
															),
														)
													);
												?>
```

sgomes replied:
Thanks, TIL!

jeherve commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:
```diff
@@ -357,10 +335,133 @@ function enqueue_assets() {
+														</p>
+													<?php else : ?>
+														<fieldset>
+															<label for="email"><?php echo( esc_html( sprintf( $required, __( 'Email', 'jetpack' ) ) ) ); ?></label>
```
```suggestion
															<label for="email"><?php echo esc_html( sprintf( $required, __( 'Email', 'jetpack' ) ) ); ?></label>
```

jeherve commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:
```diff
@@ -357,10 +335,133 @@ function enqueue_assets() {
+															<input type="text" name="email" class="jp-carousel-comment-form-field jp-carousel-comment-form-text-field" id="jp-carousel-comment-form-email-field" />
+														</fieldset>
+														<fieldset>
+															<label for="author"><?php echo( esc_html( sprintf( $required, __( 'Name', 'jetpack' ) ) ) ); ?></label>
```
```suggestion
															<label for="author"><?php echo esc_html( sprintf( $required, __( 'Name', 'jetpack' ) ) ); ?></label>
```

jeherve commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:
```diff
@@ -357,10 +335,133 @@ function enqueue_assets() {
+													name="submit"
+													class="jp-carousel-comment-form-button"
+													id="jp-carousel-comment-form-button-submit"
+													value="<?php echo( esc_attr( $localize_strings['post_comment'] ) ); ?>" />
```
```suggestion
													value="<?php echo esc_attr( $localize_strings['post_comment'] ); ?>" />
```

jeherve commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:
```diff
@@ -357,10 +335,133 @@ function enqueue_assets() {
+							</div>
+							<div class="jp-carousel-comments"></div>
+							<div id="jp-carousel-comments-loading">
+								<span><?php echo( esc_html( $localize_strings['loading_comments'] ) ); ?></span>
```
```suggestion
								<span><?php echo esc_html( $localize_strings['loading_comments'] ); ?></span>
```

jeherve commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:
```diff
@@ -357,10 +335,133 @@ function enqueue_assets() {
+					<div class="jp-carousel-image-meta">
+						<div class="jp-carousel-buttons">
+							<?php if ( $localize_strings['display_comments'] ) : ?>
+							<a class="jp-carousel-commentlink" href="#"><?php echo( esc_html( $localize_strings['comment'] ) ); ?></a>
```
```suggestion
							<a class="jp-carousel-commentlink" href="#"><?php echo esc_html( $localize_strings['comment'] ); ?></a>
```

jeherve commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:
```diff
@@ -357,10 +335,133 @@ function enqueue_assets() {
+							<a class="jp-carousel-commentlink" href="#"><?php echo( esc_html( $localize_strings['comment'] ) ); ?></a>
+							<?php endif ?>
+							<?php if ( is_user_logged_in() && $localize_strings['is_public'] && $localize_strings['reblog_enabled'] ) : ?>
+								<a class="jp-carousel-reblog" href="#"><?php echo( esc_html( $localize_strings['reblog'] ) ); ?></a>
```
```suggestion
								<a class="jp-carousel-reblog" href="#"><?php echo esc_html( $localize_strings['reblog'] ); ?></a>
```

jeherve reviewed (dismissed):
I gave it a try, it tests well. I left some comments below.

**************

In IE11, I don't seem to be able to scroll down to see the comment form when looking at an image.

sgomes commented:
Thank you for the review, @jeherve, and I apologise for making you write so many comments 😔

PHP is definitely not my forte, and while I'm more than happy to make any of the changes you suggested, it may take a while before I'm able to preemptively detect these problems on my own, before you or anyone else has to point them out.

I'll start work on those fixes right away 👍

glendaviesnz commented:
@sgomes my changes from https://github.com/Automattic/jetpack/pull/18720 have now been merged into WPcom with D57728-code so you will need to pull those changes into this PR to keep everything lined up.

sgomes commented:
Rebased to pick up changes from #18720.

sgomes commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:
```diff
@@ -357,10 +335,133 @@ function enqueue_assets() {
 			do_action( 'jp_carousel_enqueue_assets', $this->first_run, $localize_strings );
 
+			// Add the carousel skeleton to the page.
+			$this->localize_strings = $localize_strings;
```
The reason I kept this as-is is because `$localize_strings` is populated by other wpcom-specific mu-plugins, as well as provided as a parameter to at least one action. I feared removing some of the values that are now no longer directly used in JS could introduce compatibility issues.

That said, I could still create `$this->skeleton_strings` as a subset of `$localize_strings` with just the values that are needed for the skeleton, but I'm not sure how useful that would be.

sgomes commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:
```diff
@@ -357,10 +335,133 @@ function enqueue_assets() {
+							<?php if ( $localize_strings['display_comments'] ) : ?>
+							<a class="jp-carousel-commentlink" href="#"><?php echo( esc_html( $localize_strings['comment'] ) ); ?></a>
+							<?php endif ?>
+							<?php if ( is_user_logged_in() && $localize_strings['is_public'] && $localize_strings['reblog_enabled'] ) : ?>
```
It appears to be safe to do so, yes. Done.

sgomes commented:
Sorry for the delay, @jeherve, I had trouble juggling a few other things. I've gone ahead and applied all your changes and will be running some tests to make sure everything still works, between the rebase and the new commits.

sgomes commented:
I couldn't find any issues with the updated version 👍

sdixon194 commented:
Testing this one a bit. Getting this error when pulling up the carousel **when logged in**:

```
Notice: Undefined index: reblog_enabled in /srv/users/userb80fa67e/apps/userb80fa67e/public/wp-content/plugins/jetpack-dev/modules/carousel/jetpack-carousel.php on line 457
```

Shows up in the comment field on the bottom right:

![Screen Shot on 2021-03-10 at 11:26:26](https://user-images.githubusercontent.com/33553323/110663577-9206e380-8194-11eb-9430-87da5f8daabf.png)

Doesn't show for me when viewing the site when I'm not logged in however.

sgomes commented:
> Testing this one a bit. Getting this error when pulling up the carousel **when logged in**:
> 
> ```
> Notice: Undefined index: reblog_enabled in /srv/users/userb80fa67e/apps/userb80fa67e/public/wp-content/plugins/jetpack-dev/modules/carousel/jetpack-carousel.php on line 457
> ```
> 
> Shows up in the comment field on the bottom right:
> 
> ![Screen Shot on 2021-03-10 at 11:26:26](https://user-images.githubusercontent.com/33553323/110663577-9206e380-8194-11eb-9430-87da5f8daabf.png)
> 
> Doesn't show for me when viewing the site when I'm not logged in however.

Thank you, @sdixon194! 👍  That key isn't always present, unlike the others, so it needs an `in_array` check. Should be fixed with 62c806f4cf082be69a7602cf47ddb39c64101925.

sgomes commented:
Hey folks! Any chance of taking another look at this one?

sgomes commented:
Hey folks! Any chance of taking another look at this one?

sdixon194 commented:
Thanks for your patience @sgomes! I merged with the main branch again and added a changelog.

I went to do some more testing and noticed that it looks like scrolling in the gallery doesn't leave enough padding now (happened on all browsers):

**This branch:**

![Screen Capture on 2021-03-24 at 11-36-06](https://user-images.githubusercontent.com/33553323/112342212-6a397480-8c98-11eb-96b1-49312a5dc7f4.gif)


**Compared to Stable:**

![Screen Capture on 2021-03-24 at 11-35-35](https://user-images.githubusercontent.com/33553323/112341516-de274d00-8c97-11eb-97fe-211d079f38bf.gif)

I also wasn't able to scroll down at all in infamous IE11. 

Otherwise, tested with different WordPress versions, environments, etc and everything else seems okay!

sdixon194 commented:
I also just realized that the comment I left last week was never posted because it was set as Pending! I don't think it's relevant anymore, but just for the sake of documenting everything, I can't get the comment field to appear in my docker instance. Works everywhere else - regular test site with jetpack beta, JN site, etc. Just not locally for whatever reason:

![Screen Shot on 2021-03-24 at 12:31:16](https://user-images.githubusercontent.com/33553323/112346954-d7e79f80-8c9c-11eb-8876-a236c94fd54d.png)

I don't think that's a blocker, but something curious I figured was worth documenting.

sgomes commented:
Thank you for taking another look, @sdixon194! 👍 I'll be AFK for the rest of the week, but I'll pick up those styling issues as soon as I come back on Monday!

sgomes commented:
Rebased on master.

sgomes commented:
@sdixon194: Sorry for the gigantic delay in getting back to this! 😞 It was difficult to find the time to dive back into it fully between other tasks.

I've rebased it and fixed both styling issues you've mentioned.

The padding issue was fixed with a simple new rule.

The IE 11 issue took a bit more figuring out, but it's apparently caused by that browser not supporting the `overflow` shorthand with two values. Ugh. Must have been something I changed after the last IE 11 test. Splitting it up into `overflow-x` and `overflow-y` worked.

sdixon194 commented:
>Sorry for the gigantic delay in getting back to this! 😞 It was difficult to find the time to dive back into it fully between other tasks.

No worries @sgomes ! Apologies again for letting the review sit in pending for so long and pushing it back further. Is this good to review again? I'll check it out today and hopefully we can finally get it merged! 😄

sgomes commented:
> No worries @sgomes ! Apologies again for letting the review sit in pending for so long and pushing it back further.

No worries, similar things have happened to all of us 😁

> Is this good to review again? I'll check it out today and hopefully we can finally get it merged! 😄

It is, yes! I've solved the issues you'd mentioned (except for the pending comment one, no idea how to tackle that), and as far as I can tell this is working well. But you seem to have a knack for finding bugs, so I'll let you be the judge of that 🙂

sdixon194 reviewed (approved):
>  But you seem to have a knack for finding bugs, so I'll let you be the judge of that

Hahaha, well luckily, I tested this pretty extensively again and couldn't find any this time! 🎉 I think we're good - all that's left is figuring out what to do about the wpcom diff, since it was applied before the Jetpack PR 🤔

sgomes commented:
> Hahaha, well luckily, I tested this pretty extensively again and couldn't find any this time! 🎉 I think we're good - all that's left is figuring out what to do about the wpcom diff, since it was applied before the Jetpack PR 🤔

Yes, my Fusion-fu is pretty bad, so I'm not entirely sure what to do about that. Perhaps I could prepare a wpcom diff that syncs wpcom to the latest code in this PR, and we could deploy that concurrently with a forced merge of this one?

sgomes commented:
Created D60283-code to perform the sync we discussed above.

sdixon194 commented:
Great! Approved that, and was able to force the test passing. This should be good to merge! 👏

sgomes commented:
Woohoo, thank you, @sdixon194! It looks like auto-merge has been turned on (thanks, @jeherve!), so I'll keep an eye on this PR and deploy the wpcom diff when this one gets merged.

In the meantime, please let me know if I still need to take care of anything, as this is the first time I see an auto-merge!
